src/commands/scanner/run.ts

```diff
@@ -222,46 +222,71 @@ export default class Run extends ScannerCommand {
 	}
 
 	private processOutput(rrr: RecombinedRuleResults): AnyJson {
-		const {minSev, results} = rrr;
-		// If the results were an empty string, it means we found no violations.
+		const {minSev, summaryMap, results} = rrr;
+		// If the results were an empty string, then we found no violations.
 		if (results === '') {
-			// We can just log a message to the console, and also return it for the --json route.
-			const msg = messages.getMessage('output.noViolationsDetected');
+			const msg = messages.getMessage('output.noViolationsDetected', [[...summaryMap.keys()].join(', ')]);
 			this.ux.log(msg);
 			return msg;
-		} else if (this.flags.outfile) {
-			return this.writeToOutfile(minSev, results);
+		}
+		// In addition to processing the results, we need to build a summary message indicating what happened.
+		let msg = this.buildEngineSummaryMessage(rrr);
+		// We surface violations to the user by writing them to the console or a file.
+		msg += ' ' + (this.flags.outfile ? this.writeToOutfile(results) : this.writeToConsole(results));
+
+		// Now that we have our completed summary message, we need to do something with it. We'll either throw it as an
+		// exception or log it to the console.
+		if (minSev > 0 && this.flags['violations-cause-error']) {
+			throw new SfdxError(msg, null, null, minSev);
+		} else {
+			this.ux.log(msg);
+		}
+
+		// Finally, we need to return something for use by the --json flag.
+		if (this.flags.outfile) {
+			// If we used an outfile, we should just return the summary message, since that says where the file is.
+			return msg;
+		} else if (typeof results === 'string') {
+			// If the specified output format was JSON, then the results are a huge stringified JSON that we should parse
+			// before returning. Otherwise, we should just return the result string.
+			return this.determineOutputFormat() === OUTPUT_FORMAT.JSON ? JSON.parse(results) : results;
 		} else {
-			return this.formatAndDisplayOutput(minSev, results);
+			// If the results are a JSON, return the `rows` property, since that's all of the data that would be displayed
+			// in the table.
+			return results.rows;
 		}
 	}
 
 	private getInternalErrorCode(): number {
 		return this.flags['violations-cause-error'] ? INTERNAL_ERROR_CODE : 1;
 	}
 
-	private writeToOutfile(minSev: number, results: string | {columns; rows}): AnyJson {
+	private buildEngineSummaryMessage(rrr: RecombinedRuleResults): string {
+		const {summaryMap, minSev} = rrr;
+		let msg = [...summaryMap.entries()]
+			.map(([engine, summary]) => {
+				return messages.getMessage('output.engineSummaryTemplate', [engine, summary.violationCount, summary.fileCount]);
+			})
+			.join(' ');
```

Comment:
How about a newline to join so that each engine summary has its own line?

Reply:
Good idea.